memory_allocator.c

```diff
@@ -0,0 +1,258 @@
+#include <common.h>
+#include <stdint.h>
+#include <stddef.h>
+#include "queue.h"
+#include "libkern.h"
+#include "include/memory_allocator.h"
+
+/*
+  TODO:
+  - use the mp_next field of malloc_pool
+*/
+
+#define MAGIC 0xC0DECAFE
+#define ALIGNMENT 8
+
+TAILQ_HEAD(mb_list, mem_block);
+TAILQ_HEAD(mp_arena, mem_arena);
+
+static inline int32_t abs(int32_t x)
+{
+  if (x < 0)
+    return -x;
+  return x;
+}
+
+typedef struct mem_block {
+  uint32_t mb_magic; /* if overwritten report a memory corruption error */
+  int32_t mb_size; /* size < 0 => free, size > 0 => alloc'd */
+  TAILQ_ENTRY(mem_block) mb_list;
```

Comment:
Note, that there's a difference between `int b[]` and `int b[0]`. With the latter you won't get an error message and it will be happily accepted even by older compilers (i.e. like 10-15 years old if you really care about portability).